core/dbt/rpc/response_manager.py

```diff
@@ -0,0 +1,275 @@
+        self.subscriber = None
+        self.process = None
+        self.started = None
+        self.timeout = None
+        self.logs = []
+        self.task_id = uuid.uuid4()
+
+    @property
+    def request_source(self):
+        return self.http_request.remote_addr
+
+    @property
+    def request_id(self):
+        return self.json_rpc_request._id
+
+    @property
+    def method(self):
+        return self.task.METHOD_NAME
+
+    def _wait_for_results(self):
+        """Wait for results off the queue. If there is an exception raised,
+        raise an appropriate RPC exception.
+
+        This does not handle joining, but does terminate the process if it
+        timed out.
+        """
+        try:
+            msgtype, result = self.subscriber.dispatch_until_exit(
+                started=self.started,
+                timeout=self.timeout,
+            )
+        except dbt.exceptions.Exception as exc:
+            raise dbt_error(exc)
+        except Exception as exc:
+            raise server_error(exc)
+        if msgtype == QueueMessageType.Error:
+            raise RPCException.from_error(result)
+        elif msgtype == QueueMessageType.Timeout:
+            if not self.task.args.single_threaded:
+                self.process.terminate()
+            raise timeout_error(self.timeout)
+        elif msgtype == QueueMessageType.Result:
+            return result
+
+    def get_result(self):
+        try:
+            with list_handler(self.logs):
+                try:
+                    result = self._wait_for_results()
+                finally:
+                    if not self.task.args.single_threaded:
+                        self.process.join()
+        except RPCException as exc:
+            exc.logs = [l.to_dict() for l in self.logs]
+            raise
+
+        result['logs'] = [l.to_dict() for l in self.logs]
+        return result
+
+    def handle(self, kwargs):
+        self.started = time.time()
+        self.timeout = kwargs.pop('timeout', None)
+        self.subscriber = QueueSubscriber()
+        if self.task.args.single_threaded:
+            _task_bootstrap(self.task, self.subscriber.queue, kwargs)
+            return self.get_result()
+        else:
+            # this is pretty unfortunate, but we have to reset the adapter
+            # cache _before_ we fork on posix. libpq, but also any other
+            # adapters that rely on file descriptors, get really messed up if
+            # you fork(), because the fds get inherited but the state isn't
+            # shared. The child process and the parent might end up trying to
+            # do things on the same fd at the same time.
+            # Also for some reason, if you do this after forking, even without
+            # calling close(), the connection in the parent ends up throwing
+            # 'connection already closed' exceptions
+            if os.name != 'nt':
+                cleanup_connections()
```

Comment:
I'm still digging into this, but when we cleanup connections here, will that discard the relation cache?

Reply:
It shouldn't, it should only remove the actual connection handles.